packages/core/src/client.ts

```diff
@@ -197,100 +199,95 @@ export class Client {
+        return pipe(
+          result$,
+          onStart<OperationResult>(() => this.dispatchOperation(operation)),
+          take(1)
+        );
+      }
+
+      const teardown$ = pipe(
+        this.operations$,
+        filter(
+          (op: Operation) => op.kind === 'teardown' && op.key === operation.key
+        )
+      );
+
+      const refetch$ = pipe(
+        this.operations$,
+        filter(
+          (op: Operation) =>
+            op.kind === operation.kind &&
+            op.key === operation.key &&
+            (op.context.requestPolicy === 'network-only' ||
+              op.context.requestPolicy === 'cache-and-network')
+        ),
         take(1)
       );
-    }
 
-    const teardown$ = pipe(
-      this.operations$,
-      filter(
-        (op: Operation) => op.kind === 'teardown' && op.key === operation.key
-      )
-    );
+      result$ = pipe(
+        result$,
+        takeUntil(teardown$),
+        switchMap(result => {
+          if (result.stale) return fromValue(result);
+          return merge([
+            fromValue(result),
+            pipe(
+              refetch$,
+              map(() => ({ ...result, stale: true }))
+            ),
+          ]);
+        }),
+        onEnd<OperationResult>(() => {
+          active!.cache = null;
+          this.activeOperations.delete(operation.key);
+          for (let i = this.queue.length - 1; i >= 0; i--)
+            if (this.queue[i].key === operation.key) this.queue.splice(i, 1);
+          this.dispatchOperation(
+            makeOperation('teardown', operation, operation.context)
+          );
+        }),
+        onPush(result => {
+          active!.cache = result;
+        }),
+        share
+      );
 
-    const refetch$ = pipe(
-      this.operations$,
-      filter(
-        (op: Operation) =>
-          op.kind === operation.kind &&
-          op.key === operation.key &&
-          op.context.requestPolicy !== 'cache-only'
-      )
-    );
+      active = { source: result$, cache: null };
+    }
 
-    const result$ = pipe(
-      operationResults$,
-      takeUntil(teardown$),
-      switchMap(result => {
-        if (result.stale) return fromValue(result);
-
-        return merge([
-          fromValue(result),
-          pipe(
-            refetch$,
-            take(1),
-            map(() => ({ ...result, stale: true }))
-          ),
-        ]);
-      }),
+    const behavior = makeSubject<OperationResult>();
+    const cache = active.cache;
+
+    return pipe(
+      merge([behavior.source, active.source]),
       onStart<OperationResult>(() => {
-        this.onOperationStart(operation);
-      }),
-      onEnd<OperationResult>(() => {
-        this.onOperationEnd(operation);
+        const { key } = operation;
+        this.activeOperations.set(key, active!);
+        this.dispatchOperation(operation);
+        if (cache != null && cache === active!.cache)
+          behavior.next({ ...cache, stale: true });
```

Comment:
This code isn't great and I'd love to improve it so that A) it becomes part of the original source after `share` (which currently isn't possible becaue `makeSubject()` can't be shared) and B) so that the subject isn't so closely coupled to the `onStart` event's `this.dispatchOperation` call (which currently isn't possible because we can't react to any other event that comes after the potentially synchronous result)